hledger-lib/Hledger/Data/Types.hs

```diff
@@ -47,7 +47,13 @@ import Hledger.Utils.Regex
 -- | A possibly incomplete date, whose missing parts will be filled from a reference date.
 -- A numeric year, month, and day of month, or the empty string for any of these.
```

Comment:
SmartDateOrSpan ?

Reply:
I think most of the actual shrinking came from switching from monadic to applicative style in a few of the parsers, like `reportingintervalp`.

I think what the `SmartDate` type represents hasn't changed at all, just its internal representation, and it always represents an incomplete `Day` which can be transformed to a proper `Day` given a reference date. It can be incomplete either because some of the YMD are not specified, or because it's given relative to the reference date. It can also be transformed to a `DateSpan` based on some rules, but I'm not sure this is intrinsic to the SmartDate itself, as you could do the same for an ordinary Day.

That said, your description looks pretty good. I'm not sure if it should be included here, or just in the descriptions for `fixSmartDate` and `spanFromSmartDate`, respectively.

Reply:
How about `IncompleteDate` for the name of the data type? An `IncompleteDate` can be completed to either a `Day` or a `DateSpan`.

Reply:
You're right, the type's meaning hasn't changed; it felt "simpler" to me before, because of being stringly-typed.

Let's add the haddock to the type. I suggest we just leave it as SmartDate for now.

Reply:
Haddock added.